sw/device/tests/sim_dv/spi_tx_rx_test.c

```diff
@@ -107,6 +107,11 @@ void handler_irq_external(void) {
     case kTopEarlgreyPlicIrqIdSpiDeviceTxUnderflow:
       spi_device_irq = kDifSpiDeviceIrqTxUnderflow;
       CHECK(expected_irqs[spi_device_irq], "Unexpected TX underflow interrupt");
+      // TxUnderflow will fire every cycle, so disable this interrupt once fired.
```

Comment:
@eunchan @weicaiyang is this something we should fix on the design side?
i know in general we try to avoid interrupt storms

Reply:
Tx underflow was firing constantly because the SP host (the agent on the testbench) was shifting in byte after byte into the design while the `spi_device` IP had nothing to shift out (TX fifo was already emptied out in previous steps). 

This sounds like the IP is doing the right thing. This seems like a stimulus issue. Don't think there is anything to change on the design side. 

The other alternative to prevent the SPI device from firing the TX underflow beyond this point is to keep the TX fifo populated with junk data (use rand_testutils!!).

Reply:
In general if you have a situation where some behavior can storm the software with interrupts. You want to avoid it (if reasonable).  
It's reasonable for example to say once we fire an under-run interrupt to not fire it again until software has acknowledged the underlying condition in some way.

The situation you're pointing out for example, could happen if the software / hardware are properly behaving but the host has gone off the deep end.

Reply:
actually it's best to get software feedback here @cfrantz @timothytrippel @moidx 
To @sriyerg point, I don't think this is a MUST fix.

Reply:
I would think that TX underflow (and watermark even) interrupts will be enabled only when SW knows what the host is expecting to receive, and it has already populated the TX fifo. If it has no understanding of what it is the host wants (for example, the period of time when it is receiving commands from the host in TPM mode), wouldn't it keep the tx underflow disabled, because it has nothing to send (yet)? I agree SW and HW can go out of sync, but it seems like SW could be designed to avoid interrupt storms by just disabling interrupts it does not care about at a certain time. 

I would think this will apply to GPIOs as well (or pretty much anything that can cause an interrupt externally) - if ithe SW enabled all GPIO interrupts, the external world could technically randomly keep wiggling pins and create an interrupt storm. We can easily create that scenario in simulation, but will something like this actually occur in production. Seems unlikely. Not sure this is something the hardware design needs to (or can) protect against. 

I would also think that main flow of execution being severely starved like this is something that can be detected with the watchdog timer, but not sure of that either.

Reply:
so for the tx underflow, in general it should not happen. So the way I have seen it used at least, is that the software actually keeps the interrupt enabled, and when it fires, treat it as some kind of serious error that requires triage. Because at that point communication between the device / host can no longer be trusted. And while it's true they could just disable it when it happens... let's just say I've had a few stops where some interrupts that require silencing and others being event driven (when the event is held) kind of causes confusion in software handling. 

And you are right GPIO constantly toggling would also cause an issue, but for peripheral interfaces I think it's more likely for the host for example to get wedged and either a) not toggle clocks at all, or b) keep them toggling indefinitely for no reason.  The not toggle at all I think is part of the reason blocks like i2c have some kind of error detection to know that host behavior is completely wrong and requires some kind of higher level handling. 

This is why I added to my statement that IF we can do something to easily avoid an interrupt storm in hardware, we should try, assuming the effort is reasonable. It is also why I don't view this as a MUST fix. 

It would be good to get software's opinion, because depending on how interrupts are handled, specific ones that need to be disabled vs others where we just need to clear the state and handle later don't always make a huge difference.  I have for example seem systems that have a front-end / back-end processing to interrupts. So in the front end they want to do something as generic as possible, and prefer to keep out things like IF SPI & INTERRUPT_X, go disable elsewhere.

Reply:
btw i should add, all your points are valid, i'm just raising some counter-examples.
It would be good to see how the software folks feel.

Reply:
Sounds good, thanks! This is all interesting to me. Looking fwd to hearing from SW team on this.

Reply:
I think this PR can go in, given this is discussion can happen on the side. WDYT?

Reply:
Thanks for raising up the potential interrupt storm issue. Would like to know if handling this on SW side is a concern from the SW prospective.

I will merge PR once CI passes and we can continue the discussion.

Reply:
I think I agree here that an IRQ like TX underflow would only be enabled when on-device SW begins a TX sequence (to the host). But I'm also not sure I follow the scenario where software would leave the underflow IRQ enabled if it's not trying to transmit anything? Are you thinking of a super-loop-type SW arch. scenario where the TX underflow IRQ fires, and the SW IRQ handler wants to just set some type of global flag saying "hey dump more data in the TX fifo when you get a chance", clears the interrupt, and then in some main control loop actually dumps the data to TX fifo (or instructs some DMA controller to do so)? And, the only way to keep the TX underflow IRQ from continuously firing is to keep disabling/re-enabling it? Maybe you can elaborate more offline. Sorry for any misunderstanding here!

Reply:
yeah i'm not quite sure what the right term is, but I've seen a few cases where the interrupt handling is not immediately done. Instead for latency purposes, software mostly "remembers" the fact that an interrupt fired, and files it away for handling in a later loop.  So from what I recall in these scenarios, there's a desire to just quickly acknowledge the interrupts in a generic way and move on.  I cant' recall now in these architectures whether the preferred way is to silence the interrupt in the plic type structure, or just clear the state. I want to say the latter, because I have very specific recollection that a level type event (or something that just repeatedly triggers) broke this particular software assumption. 

I think specifically to underrun, my point was more, ONCE it happens, we already know something is irreversibly broken.  There really is no need to keep on firing underrun interrupts (it has gotten MORE broken!) until software handles the event or acknowledges it.  Does that make sense? 

In a scenario where we are doing transfers with the host, but the host somehow goes haywire and just keeps toggling for no reason, we'd keep getting these interrupts, whereas the likely system behavior is that if we get this underrun interrupt, we probably need to resynchronize with the host and restart that link.

Reply:
Ahhh yes, this was exactly the scenario I was trying to describe above too, glad I'm on the same page! This type of SW architecture keeps your system more responsive (i.e., [keeping ISRs as short as possible keeps the system responsive](https://betterembsw.blogspot.com/2013/04/why-short-interrupt-service-routines.html)), and is typically considered good practice.

Regarding the underrun this makes much more sense, thanks for clarifying. From a SW perspective, it would be easier to address the event that caused the IRQ if it is not constantly re-fired. Is the idea here that the HW would be altered to fire underrun IRQ only once, and only reset the interrupt to be able to fire again once the link has been restarted or the fifo has been filled? Obviously we can work around this issue by just disabling/re-enabling IRQs, as mentioned above, but it seems your approach would certainly make things more intuitive from the SW side.

Am curious to hear others' thoughts as well on this!

Reply:
Thanks for the info. I agree if an event occurs and SW does not clear the cause of the event, HW would be better to notify SW once. 

In SPI_DEVICE, the tx underflow condition occurs at every byte when FIFO is empty. The logic does not constantly fire the interrupts at every cycle but every 8 SCK cycles. So, to mute the interrupt event, I think logic needs to be change to check the ASYNC_FIFO read depth too and does not assert twice until the read depth becomes non-zero.

Does it sound good to everyone? Then I can create a PR to change the interrupt behavior (maybe RX overflow too?)